### ISSUE TITLE
Need to fix NSM behavior with missing optional, named and type arguments

### ISSUE THREAD

**Issue by sjindel-google** [dart-lang/sdk#32794]
Our current behavior when an NSM forwarder is called with missing optional, named or type arguments is to preserve their absence in the `Invocation` object passed to `noSuchMethod`. However, the spec. says they should be filled in with their default values.

In this example:

```dart
class C {
  dynamic noSuchMethod(Invocation invoke) {
    print(invoke.positionalArguments);
    print(invoke.namedArguments);
    print(invoke.typeArguments);
  }

  void f1<T>();
  void f2([int x = 3]);
  void f3({int y = 3});
}

void main() {
  var c = new C();
  c.f1();
  c.f2();
  c.f3();
}
```

Our behavior currently is:

```
[]
{}
[]
[]
{}
[]
[]
{}
[]
```

Whereas it should be:
```
[]
{}
[dynamic]
[3]
{}
[]
[]
{Symbol("y"): 3}
[]
```

**Comment by sjindel-google** [comment]
@leafpetersen: Please adjust the priority based on your understand of the impact fixing this behavior will have on "Bazel" and Flutter.

**Comment by eernstg** [comment]
It may be surprising (and it was [discussed](https://dart-review.googlesource.com/c/sdk/+/6500/7/docs/language/informal/nosuchmethod-forwarding.md#216), starting from the assumption that we would "obviously" want to have the actual argument list passed in the `Invocation`, such that arguments not passed would be absent, not defaulted). But there was a clear decision that the forwarders should pass an `Invocation` to nSM that specifies the bindings of the formal parameters, no matter which actual argument list was used to create those bindings. We also noted that this approach cannot be used for a dynamic invocation where no forwarder matches: There is simply no formal parameter list available.

One of the reasons for selecting this semantics was that we did not want to make nSM + forwarder a "more powerful version" of a regular method for normal, typed programming, that is, we did not want to encourage developers to use nSM a lot, as a device for treating default parameters in a more fancy manner than they could have done with a regular method.

So it may be a wart on the language that dynamic and statically checked method invocations have different semantics in this respect, but it is the decision that was made, after considerable debate.

**Comment by stefantsov** [comment]
Interesting.  I remember some line in the following test was also checking that the actual parameters are passed to noSuchMethod, not the default values: https://dart.googlesource.com/sdk/+/master/tests/language_2/no_such_method_mock_test.dart.  Anyways, that's probably a bug in the test.

Great!  Now that we know that there's no magic behind the forwarders, I can implement them fully in fasta, and the back ends may just consume normal Kernel code.

**Comment by eernstg** [comment]
Cool! And I agree that [the test](https://dart.googlesource.com/sdk/+/master/tests/language_2/no_such_method_mock_test.dart) must be adjusted in line 85 and 88 (and possible elsewhere, but those two jumped out at me ;-).

**Comment by leafpetersen** [comment]
Agreed. Priority is very low.  Vanishingly unlikely that there is any code relying on the current (incorrect) DDC behavior or any behavior for this at all.

**Comment by sjindel-google** [comment]
> Agreed. Priority is very low.  Vanishingly unlikely that there is any code relying on the current (incorrect) DDC behavior or any behavior for this at all.

It turns out this breaks Mockito severely: #33031

**Comment by eernstg** [comment]
@lrhn and I discussed this, and @lrhn later wrote up some arguments [here](https://github.com/dart-lang/mockito/issues/122#issuecomment-386608641) supporting the claims that (1) Mockito should just keep working in a number of situations (in particular, when actual argument values are passed rather than the more "abstract" matchers), (2) it should be possible (and presumably not terribly difficult) to update Mockito to handle the newly provided default arguments gracefully, and (3) the only behaviors that Mockito won't be able to mock are behaviors that no actual Dart method can implement anyway, so it shouldn't be a problem that we cannot mock them any more.

**Comment by dgrove** [comment]
Is this still blocked?

**Comment by sjindel-google** [comment]
Completing the task entails removing VM support, but I would like to wait until the new semantics have been successfully rolled into Flutter before continuing.

**Comment by dgrove** [comment]
Since this is P3,  should this be moved from the milestone?
